### DISCUSSION TITLE
[Issue] Connect to Excel or Sharepoint Data Sources?

### DISCUSSION THREAD

**Discussion by MattGrove36** [TabularEditor/TabularEditor3#1018] [🙋♀️TE3 Q&A] [answered]
Hi-
I'm trying to connect to non-SQL data sources, specifically Excel workbooks and SharePoint on-prem and SharePoint online. I have PowerQuery (M) code which works fine when trying to connect to those sources in PowerBI or Excel. However when I try to setup a Data Source or import to a table, I am unsuccessful and get this error message: Unknown error while parsing M expression: Object reference not set to an instance of an object

I am using Tabular Editor 3, version 3.10.1 Business edition. Windows 10 Pro.

Thank you

**Comment by edhans** [comment, marked as answer]
Just right-click on Tables in the TOM, and select Create, then Table. It should create a table with an M partition. Drop your entire M query into the partition M Expression field.

I don't think you can do this via the Import method, though I'll be curious to see if the TE team shows a way you can.

**Comment by MattGrove36** [comment]
Wow, I don't know why this simple solution eluded me :-) I tried just about every other thing I could think of. Thank you!!

**Comment by MattGrove36** [comment]
Maybe I rejoiced too soon. This method was successful for connecting to a single Excel workbook as a data source. It doesn't work for connecting to SharePoint nor for connecting to an entire folder full of files. When attempting either of those, now I get this error message: 

"An M partition uses a data function which results in access to a data source different from those defined in the model"

Any suggestions? I am using Tabular Editor 3, version 3.10.1 Business edition. Windows 10 Pro.

**Comment by edhans** [comment]
When connecting to a folder, the combine operation will create 4 shared expressions you will need to manually copy the code over. When you create the shared expression, you must tell TE to make it an M kind. It won't do that automatically. It is in the properties section of the shared expression.

<img width="135" alt="image" src="https://github.com/TabularEditor/TabularEditor3/assets/45298358/5925c3dc-abd2-474a-a09f-24a755d7414a">

They must be named _exactly_ the same way, even the (1) or (2) at the end of you have multiple combine operations. I've never added a folder combine operation like this in TE, but I don't know why it wouldn't work. Just tedious to set up.
